compiler/src/dotty/tools/dotc/transform/PostTyper.scala

```diff
@@ -415,6 +407,17 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
         case tree: New if isCheckable(tree) =>
           Checking.checkInstantiable(tree.tpe, tree.srcPos)
           super.transform(tree)
+        case cdef @ CaseDef(pat, guard, tree @ AssumeInfo(sym, nestingLevel, info, body)) =>
+          // test case pos/i9833
+          val (assumeInfos, body) = tree.unnest()
+          inContext(ctx.withGadt(ctx.gadt.withAssumeInfos(assumeInfos))) {
+            val pat1   = transform(pat)
+            val guard1 = transform(guard)
```

Comment:
This is necessary to satisfy the bounds check that happens in post typer.  On the opposite side it needs to especially not be done in the TypeTestCasts checks, which are run during Erasure, in order to give correct results on whether any type test has any uncheckable parts.

It comes down to the fact that the assumed info may only be assume _provided_ the pattern matches.  So in PostTyper we want to check the bounds, which only matters if the pattern matches, while in TypeTestCasts we want to check uncheckable, which may not assume the pattern matches.

Wrapping the CaseDef is double hard.  Firstly, because undoing the assumption for TypeTestCasts isn't easy.  Secondly, because Match is already reused for inline matches and match types, so having all those code paths deal with this change is painful.  There's a lot of code that expects matches to be of cases, so it's quite hairy to make it all deal with AssumeInfo vs CaseDef.  In my original where I was holding onto a GadtConstraint I tried splitting CaseDef into {GadtCaseDef, CaseDef} <: GenericCaseDef, but it was very painful.  I also tried having all CaseDef have a GadtConstraint, but it was still awkward.

Repeating the required AssumeInfo still runs into the TypeTestCasts issue.  And would that be repeating as in copy or same reference?  I could reuse the same reference, and do something awful for TypeTestCasts.

Btw, if you look at the before state (earlier in the file) it was set up in the opposite direction: the gadt as an attachment on the pattern was installed to be present when transforming the tree as a whole.  So this is different but the same, as I see it.